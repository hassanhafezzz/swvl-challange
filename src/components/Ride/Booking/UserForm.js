import React, { useContext } from 'react';
import { Formik, Form } from 'formik';
import classNames from 'classnames/bind';
import { addBooking } from '../../../store/actions';
import { Context } from '../../../store';
import styles from './styles.module.css';
import InputField from '../../common/Form/InputField';
import SelectField from '../../common/Form/SelectField';
import RadioField from '../../common/Form/RadioField';
import Button, { BUTTON_VARIANT } from '../../common/Button';
import usersMock from '../../../data/users';
import { isLink } from '../../../utils';
import route from '../../../data/route';

const cx = classNames.bind(styles);

const stations = route.map((point) => point.name);

const UserForm = ({ closeModalForm }) => {
  const user = usersMock[Math.floor(Math.random() * usersMock.length)];

  const initialValues = {
    id: user.id,
    name: user.name,
    image: user.image,
    tripsCount: user.trips_count,
    pickupStation: '',
    dropOffStation: '',
    paymentMethod: 'cash',
  };

  const [_, dispatch] = useContext(Context);

  const onSubmit = (values) => {
    dispatch(addBooking(values));
    closeModalForm();
  };

  const validate = (values) => {
    const errors = {};
    if (!values.name) {
      errors.name = 'Who is booking the ticket?';
    }

    if (!values.pickupStation) {
      errors.pickupStation = 'Where should we pick him?';
    }

    if (!values.dropOffStation) {
      errors.dropOffStation = 'Where should we drop him to?';
    }

    if (!values.image) {
      errors.image = 'If you could provide an image that would be great';
    } else if (!isLink(values.image)) {
      errors.image = "That's not a valid link";
    }

    return errors;
  };

  const pickupStations = stations.filter((s, i) => i !== stations.length - 1);

  return (
    <Formik
      initialValues={initialValues}
      validate={validate}
      onSubmit={onSubmit}
    >
      {({ isValid, values, setFieldValue }) => {
        const dropOffStations = stations.slice(
          stations.indexOf(values.pickupStation) + 1,
        );

        return (
          <Form>
            <InputField type="text" name="name" id="name" label="name" />
            <InputField type="url" name="image" id="image" label="image" />

            <SelectField
              name="pickupStation"
              id="pickupStation"
              label="pick up station"
              onChange={(e) => {
                setFieldValue('pickupStation', e.target.value);
                setFieldValue('dropOffStation', '');
              }}
            >
              <option
                disabled
                selected
                label="Please select a pick up station"
              />
              {pickupStations.map((station) => (
                <option key={station} value={station} label={station} />
              ))}
            </SelectField>

            <SelectField
              name="dropOffStation"
              id="dropOffStation"
              label="drop off station"
            >
              <option disabled selected label="Please select a drop off" />
              {dropOffStations.map((station) => (
                <option key={station} value={station} label={station} />
              ))}
            </SelectField>

            <p className={cx('form-label')}>payment method</p>
            <RadioField
              label="cash"
              name="paymentMethod"
              id="cash"
              checked={values.paymentMethod === 'cash'}
              value="cash"
            />
            <RadioField
              label="credit"
              name="paymentMethod"
              id="credit"
              checked={values.paymentMethod === 'credit'}
              value="credit"
            />
            <Button
              variant={BUTTON_VARIANT.SECONDARY}
              disabled={!isValid}
              type="submit"
            >
              Add
            </Button>
          </Form>
        );
      }}
    </Formik>
  );
};

export default UserForm;
