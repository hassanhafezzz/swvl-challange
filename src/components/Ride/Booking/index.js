import React, { useState } from 'react';
import classNames from 'classnames/bind';
import styles from './styles.module.css';
import Avatar from '../../common/Avatar';
import Button, { BUTTON_VARIANT } from '../../common/Button';
import Modal from '../../common/Modal';
import driver from '../../../img/driver.jpg';

const cx = classNames.bind(styles);

const Booking = () => {
  const [isModalOpen, setModalOpen] = useState(false);
  const toggleModal = () => setModalOpen(!isModalOpen);
  const closeModal = () => setModalOpen(false);

  return (
    <>
      <Modal
        title="Add a new booking 👨🏻‍💻"
        isOpen={isModalOpen}
        closeModal={closeModal}
        actionText="Add"
        action={() => console.log('TODO //')}
      >
        placeholder for form
      </Modal>
      <div className={cx('heading')}>
        <h3 className={cx('title')}>
          Bookings{' '}
          <span role="img" aria-label="bus">
            🔖
          </span>
        </h3>
        <Button variant={BUTTON_VARIANT.PRIMARY} onClick={toggleModal}>
          New Book
        </Button>
      </div>
      <div className={cx('wrapper')}>
        <Avatar img={driver} name="hi" tripsNumber={12} />
      </div>
    </>
  );
};

export default Booking;
