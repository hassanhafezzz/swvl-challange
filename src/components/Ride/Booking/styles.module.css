@value utils: '../../../styles/utils.css';
@value large-bp from utils;
@value small-bp from utils;

.container {
  margin: 25px 0 10px;
  padding-bottom: 50px;
}

.heading {
  margin: 5px 0 10px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.title {
  margin: 5px 0;
  font-size: var(--h1-size);
}

.wrapper {
  margin: 25px 0;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 30px 20px;
}

.fill-bookings-container {
  margin-top: 20px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.fill-bookings-container p {
  font-size: var(--h4-size);
  font-weight: var(--bold);
}

.form-label {
  margin: 20px 0 0;
  display: flex;
  text-transform: capitalize;
  font-size: var(--small-paragraph);
  font-weight: var(--bold);
}

@media (large-bp) {
  .wrapper {
    grid-template-columns: repeat(2, 1fr);
  }
}

@media (small-bp) {
  .wrapper {
    grid-template-columns: repeat(1, 1fr);
  }
}
