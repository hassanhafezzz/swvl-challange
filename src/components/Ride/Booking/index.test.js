import React from 'react';
import { render } from '@testing-library/react';
import StoreProvider from '../../../store';
import Booking from './index';

const ele = (
  <StoreProvider>
    <Booking />
  </StoreProvider>
);

describe('Booking component test', () => {
  test('renders booking section', () => {
    const { getByTestId } = render(ele);
    const booking = getByTestId('booking');
    expect(booking).toBeTruthy();
  });

  test('add new button exists', () => {
    const { getByText } = render(ele);
    expect(getByText(/new book/i)).toBeTruthy();
  });

  test('make sure booking buttons are shown', () => {
    const { container } = render(ele);
    const buttons = container.querySelectorAll('button');
    expect(buttons.length).toEqual(2);
  });

  test('add snapshot for booking', () => {
    const { container, asFragment } = render(ele);
    expect(asFragment(container)).toMatchSnapshot();
  });
});
