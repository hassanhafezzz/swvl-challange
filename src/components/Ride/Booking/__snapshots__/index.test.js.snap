// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking component test add snapshot for booking 1`] = `
<DocumentFragment>
  <section
    class="container"
    data-testid="booking"
  >
    <div
      class="heading"
    >
      <h3
        class="title"
      >
        Bookings 
        <span
          aria-label="bus"
          role="img"
        >
          ðŸ”–
        </span>
      </h3>
      <button
        class="button secondary"
        type="button"
      >
        New Book
      </button>
    </div>
    <div
      class="fill-bookings-container"
    >
      <p>
        <span
          aria-label="sparks"
          role="img"
        >
          âœ¨
        </span>
        Or just fill all 12 spots at once 
        <span
          aria-label="sparks"
          role="img"
        >
          âœ¨
        </span>
      </p>
      <button
        class="button primary"
        type="button"
      >
        fill all
      </button>
    </div>
  </section>
</DocumentFragment>
`;
